firedrake/projection.py

```diff
@@ -101,6 +102,65 @@ def project(v, V, bcs=None, mesh=None,
     return ret
 
 
+def reconstruct(v_b, V, name=None):
+    """Reconstruct a :class:`.Function`, defined on a broken function
+    space and transfer its data into a function defined on an unbroken
+    finite element space.
+
+    In other words: suppose we have a function v defined on a space constructed
+    from a :class:`ufl.BrokenElement`. This methods allows one to "project"
+    the data into an unbroken function space.
+
+    This method avoids assembling a mass matrix system to solve a Galerkin
+    projection problem; instead kernels are generated which computes weighted
+    averages between facet degrees of freedom.
+
+    :arg v_b: the :class:`.Function` to reconstruct.
+    :arg V: the :class:`.FunctionSpace` or :class:`.Function` to reconstruct
+            into.
+    :arg name: name of the resulting :class:`.Function`
+    """
+
+    if not isinstance(v_b, function.Function):
+        raise RuntimeError("Argument must be a function. Not %s" % type(v_b))
+
+    if not isinstance(v_b.function_space().ufl_element(), ufl.BrokenElement):
+        raise ValueError("Function space must be defined on a broken element.")
+
+    if isinstance(V, functionspaceimpl.WithGeometry):
+        result = function.Function(V, name=name)
+    elif isinstance(V, function.Function):
+        result = V
+        V = V.function_space()
+    else:
+        raise RuntimeError(
+            'Can only project into functions and function spaces, not %r'
+            % type(V))
+
+    if v_b.function_space().ufl_element()._element != V.ufl_element():
+        raise ValueError(
+            "The ufl element of the target function space must "
+            "coincide with the element broken by ufl.BrokenElement."
+        )
+
+    weight_kernel = """
+    for (int i=0; i<weight.dofs; ++i) {
+    weight[i][0] += 1.0;
+    }"""
+
+    average_kernel = """
+    for (int i=0; i<vrec.dofs; ++i) {
+    vrec[i][0] += v_b[i][0]/weight[i][0];
+    }"""
+
+    w = function.Function(V)
+    par_loop(weight_kernel, ufl.dx, {"weight": (w, INC)})
+    par_loop(average_kernel, ufl.dx, {"vrec": (result, INC),
+                                      "v_b": (v_b, READ),
+                                      "weight": (w, READ)})
+    return result
```

Comment:
Something like `result.dat /= w.dat` or `result.dat.data[:] /= w.dat.data_ro[:]`.  The latter is a NumPy operation.  Not sure how the former is implemented.